linker/resolve.go

```diff
@@ -76,13 +37,50 @@ func (r *result) resolveElement(name protoreflect.FullName) protoreflect.Descrip
 	if len(name) > 0 && name[0] == '.' {
 		name = name[1:]
 	}
-	importedFd, res := resolveElement(r, name, false, nil)
-	if importedFd != nil {
-		r.markUsed(importedFd.Path())
-	}
+	res, _ := resolveInFile(r, false, nil, func(f File) (protoreflect.Descriptor, error) {
+		d := resolveElementInFile(name, f)
+		if d != nil {
+			return d, nil
+		}
+		return nil, protoregistry.NotFound
+	})
 	return res
 }
 
+func resolveInFile[T any](f File, publicImportsOnly bool, checked []string, fn func(File) (T, error)) (T, error) {
+	var zero T
+	path := f.Path()
+	for _, str := range checked {
+		if str == path {
+			// already checked
+			return zero, protoregistry.NotFound
+		}
+	}
+	checked = append(checked, path)
+
+	res, err := fn(f)
+	if err == nil {
+		return res, nil
+	}
+	imports := f.Imports()
+	for i, l := 0, imports.Len(); i < l; i++ {
+		imp := imports.Get(i)
+		if publicImportsOnly && !imp.IsPublic {
+			continue
+		}
+		res, err := resolveInFile(f.FindImportByPath(imp.Path()), true, checked, fn)
+		if err == nil {
```

Comment:
Do we need to handle `err` here if non-nil?

Reply:
Good question. Let me reformulate a little to make this more intuitive and robust.

Reply:
I've re-worked the error handling a bit. The main benefit is better error messages when a name is the wrong type (like trying to resolve a message name, and the name is found but is not a message).